File: mk.clj
(ns mk
  (:require [clojure.string :as string])
  (:refer-clojure :exclude [== conj disj cons list list? take]))

;;; First, we need a real cons cell implementation:

(defprotocol ICons
  (car [this])
  (cdr [this]))

(deftype Cons [a d]
  ICons
  (car [this] a)
  (cdr [this] d)

  Object
  (toString [this]
    (if (proper-list? this)
      (str "("
           (string/join " " (map #(if (nil? %) "nil" (.toString %)) this)) ")")
      (str "(" a " . " d ")")))

  clojure.lang.Seqable
  (seq [this] (seq-cons-list this)))

(defn cons [a d]
  (Cons. a d))

(defn proper-list? [cell]
  (or (nil? cell)
      (and (list? cell)
           (proper-list? (cdr cell)))))

(defn seq-cons-list [cell]
  (when-not (nil? cell)
    (clojure.core/cons (car cell) (cdr cell))))

(defn to-cons-list [l]
  (when (seq l)
    (cons (first l) (to-cons-list (rest l)))))

(defn list [& l]
  (to-cons-list l))

(defn list? [v]
  (instance? Cons v))



;;; µKanren begins here:

(defn lvar [c] #{c})
(defn lvar? [x] (set? x))

(defn walk [u s]
  (if (and (lvar? u) (contains? s u))
    (walk (s u) s)
    u))

(def empty-state [{} 0])

(def mzero nil)
(defn unit [sc] (cons sc mzero))

(defn unify [u v s]
  (let [u (walk u s)
        v (walk v s)]
    (cond
      (and (lvar? u) (lvar? v) (= u v)) s

      (lvar? u) (assoc s u v)

      (lvar? v) (assoc s v u)

      (and (list? u) (list? v))
      (let [s (unify (car u) (car v) s)]
        (and s (unify (cdr u) (cdr v) s)))

      :else (and (= u v) s))))

(defn == [u v]
  (fn [[s c]]
    (let [s (unify u v s)]
      (if s (unit [s c]) mzero))))

(defn callfresh [f]
  (fn [[s c]]
    ((f (lvar c)) [s (+ c 1)])))

(defn mplus [$1 $2]
  (cond
    (fn? $1) (fn [] (mplus $2 ($1)))

    (list? $1)
    (cons (car $1) (mplus (cdr $1) $2))

    :else $2))

(defn bind [$ g]
  (cond
    (fn? $) (fn [] (bind ($) g))

    (list? $)
    (mplus (g (car $)) (bind (cdr $) g))

    :else mzero))

(defn disj [g1 g2] (fn [sc] (mplus (g1 sc) (g2 sc))))
(defn conj [g1 g2] (fn [sc] (bind (g1 sc) g2)))

(defn callgoal [g]
  (g empty-state))

(defn pull [$]
  (if (fn? $) (pull ($)) $))

(defn take [n $]
  (if (zero? n) nil
      (let [$ (pull $)]
        (if (nil? $) nil (cons (car $) (take (- n 1) (cdr $)))))))

(defn take-all [$]
  (let [$ (pull $)]
    (if (nil? $) nil (cons (car $) (take-all (cdr $))))))

(defmacro fresh [vars & body]
  `(callfresh
    (fn [~(first vars)]
      ~(if (seq (rest vars))
         `(fresh ~(rest vars) (conj+ ~@body))
         `(conj+ ~@body)))))

(defn zzz [g] (fn [sc] (fn [] (g sc))))

(defn walk* [v s]
  (let [v (walk v s)]
    (cond
      (lvar? v) v

      (list? v)
      (cons (walk* (car v) s) (walk* (cdr v) s))

      :else v)))

(defn reify-1st [[s c]]
  (walk* (lvar 0) s))

(defn run [n g]
  (map reify-1st (take n (callgoal g))))

(defn run* [g]
  (map reify-1st (take-all (callgoal g))))



(defmacro zzz [g]
  `(fn [sc#] (fn [] (~g sc#))))

(defmacro conj+ [g0 & gs]
  (if (seq gs)
    `(conj (zzz ~g0) (conj+ ~@gs))
    `(zzz ~g0)))

(defmacro disj+ [g0 & gs]
  (if (seq gs)
    `(disj (zzz ~g0) (disj+ ~@gs))
    `(zzz ~g0)))

(defmacro conde [& gs]
  `(disj+ ~@(map (fn [l] `(conj+ ~@l)) gs)))



;; Infinite fives and sixes goal:

(defn fives [x]
  (disj (== x 5) (zzz (fives x))))
(defn sixes [x]
  (disj (== x 6) (zzz (sixes x))))

(def fives-and-sixes (callfresh (fn [x] (disj (fives x) (sixes x)))))



;; And finally, appendo:

;; (defn appendo [l r out]
;;   (disj
;;    (conj (== l nil) (== r out))
;;    (fresh [a d res]
;;      (== (cons a d) l)
;;      (== (cons a res) out)
;;      (zzz (appendo d r res)))))

;; (defn appendo [l r out]
;;   (disj+
;;    (conj+ (== l nil) (== r out))
;;    (fresh [a d res]
;;      (== (cons a d) l)
;;      (== (cons a res) out)
;;      (appendo d r res))))

(defn appendo [l r out]
  (conde
   ((== nil l) (== r out))
   ((fresh [a d res]
      (== (cons a d) l)
      (== (cons a res) out)
      (appendo d r res)))))

;; (run* (fresh [q a b] (== q (list a b)) (appendo a b (list 1 2 3 4 5))))

File: mk.js
import { Map, Iterable } from "immutable";

const runThunk = (v) => typeof v === "function" ? runThunk(v()) : v;

const nil = "Nil";

class Var {
  constructor(c) { this.var = c; }
  equals(o) { return this.var === o.var; }
  hashCode() { return this.var; }
  toString() { return "?" + this.var; }
}
const mkvar = (c) => new Var(c);
const isvar = (c) => c instanceof Var;

class Stream {
  constructor(a, d) { this.a = a; this.d = d; }
  *[Symbol.iterator]() {
    yield this.car();
    const next = this.cdr();
    if (next !== nil) yield* next;
  }
  car() { return runThunk(this.a); }
  cdr() { return runThunk(this.d); }
  toArray() { return toJSArray(this); }
  toString() {
    return properList(this) ? "[" + this.toArray().join(", ") + "]" :
      "(" + this.car() + " . " +
      this.cdr() + ")";
  }
}
const cons = (car, cdr) => new Stream(car, cdr);
const listp = (s) => s instanceof Stream;
const list = (vs) => vs.length ? cons(vs[0], list(vs.slice(1))) : nil;
const properList = (l) => {
  if (l === nil) { return true; }
  if (!listp(l.cdr()) && l.cdr() !== nil) { return false; }
  return properList(l.cdr());
};
const toJSArray = (l) => l === nil ? [] : [l.car()].concat(toJSArray(l.cdr()));

const walk = (u, s) => isvar(u) && s.has(u) ? walk(s.get(u), s) : u;

const emptystate = [Map(), 0];

const mzero = nil;
const unit = (x) => cons(x, nil);

const unify = (u, v, s) => {
  if (s === nil) { return nil; }
  if (u instanceof Array) { u = list(u); }
  if (v instanceof Array) { v = list(v); }
  u = walk(u, s); v = walk(v, s);
  if (isvar(u) && isvar(v) && u.equals(v)) { return s; }
  if (isvar(u)) { return s.set(u, v); }
  if (isvar(v)) { return s.set(v, u); }
  if (listp(u) && listp(v)) {
    return unify(u.cdr(), v.cdr(), unify(u.car(), v.car(), s));
  }
  if (u === v) { return s; }
  return nil;
};

const eq = (u, v) => ([s, c]) => {
  const sp = unify(u, v, s);
  return (sp === nil) ? mzero : unit([sp, c]);
};

const callFresh = (f) => ([s, c]) => f(mkvar(c))([s, c + 1]);

const mplus = (s1, s2) => {
  s1 = runThunk(s1);
  if (s1 === nil) { return s2; }
  return cons(s1.car(), () => mplus(s2, s1.cdr()));
};

const bind = (s, g) => {
  if (s === nil) { return mzero; }
  if (typeof s === "function") { return bind(s(), g); }
  return mplus(g(s.car()), () => bind(s.cdr(), g));
};

const disj = (g1, g2) => (sc) => mplus(g1(sc), g2(sc));
const conj = (g1, g2) => (sc) => bind(g1(sc), g2);

const walk1 = (v, s) => {
  const c = walk(v, s);
  if (listp(c)) { return cons(walk1(c.car(), s), walk1(c.cdr(), s)); }
  return c;
};
const reify1st = ([s, c]) => walk1(mkvar(0), s);

const callGoal = (g) => g(emptystate);
const pull = (n, g) => {
  const r = runThunk(callGoal(g));
  return r === nil ? Iterable() : Iterable(r[Symbol.iterator]()).take(n).map(reify1st);
};

const appendo = (l, r, out) => disj(
  conj(eq(l, nil), eq(r, out)),
  callFresh((a) =>
  callFresh((d) =>
  callFresh((res) =>
            conj(eq(cons(a, d), l),
                 conj(eq(cons(a, res), out),
                      (sc) => () => appendo(d, r, res)(sc))))))
);

// const fives = (x) => disj(eq(x, 5), (sc) => () => fives(x)(sc));
// const sixes = (x) => disj(eq(x, 6), (sc) => () => sixes(x)(sc));
// const fivesAndSixes = callFresh((x) => disj(fives(x), sixes(x)));
// console.log(pull(10, fivesAndSixes));

// const wat = callFresh((q) =>
//             callFresh((a) =>
//             callFresh((b) =>
//                       conj(eq(q, [a, b]),
//                            appendo(a, b, [1, 2, 3, 4, 5])))));
// console.log(pull(10, wat));
